src/platform_impl/windows/keyboard_layout.rs

```diff
@@ -0,0 +1,696 @@
+        let lalt = key_state[winuser::VK_LMENU as usize] & 0x80 != 0;
+        let ralt = key_state[winuser::VK_RMENU as usize] & 0x80 != 0;
+
+        let caps = key_state[winuser::VK_CAPITAL as usize] & 0x01 != 0;
+
+        let mut result = WindowsModifiers::empty();
+        if shift || lshift || rshift {
+            result.insert(WindowsModifiers::SHIFT);
+        }
+        if control || lcontrol || rcontrol {
+            result.insert(WindowsModifiers::CONTROL);
+        }
+        if alt || lalt || ralt {
+            result.insert(WindowsModifiers::ALT);
+        }
+        if caps {
+            result.insert(WindowsModifiers::CAPS_LOCK);
+        }
+
+        result
+    }
+
+    pub fn apply_to_kbd_state(self, key_state: &mut [u8; 256]) {
+        if self.intersects(Self::SHIFT) {
+            key_state[winuser::VK_SHIFT as usize] |= 0x80;
+        } else {
+            key_state[winuser::VK_SHIFT as usize] &= !0x80;
+            key_state[winuser::VK_LSHIFT as usize] &= !0x80;
+            key_state[winuser::VK_RSHIFT as usize] &= !0x80;
+        }
+        if self.intersects(Self::CONTROL) {
+            key_state[winuser::VK_CONTROL as usize] |= 0x80;
+        } else {
+            key_state[winuser::VK_CONTROL as usize] &= !0x80;
+            key_state[winuser::VK_LCONTROL as usize] &= !0x80;
+            key_state[winuser::VK_RCONTROL as usize] &= !0x80;
+        }
+        if self.intersects(Self::ALT) {
+            key_state[winuser::VK_MENU as usize] |= 0x80;
+        } else {
+            key_state[winuser::VK_MENU as usize] &= !0x80;
+            key_state[winuser::VK_LMENU as usize] &= !0x80;
+            key_state[winuser::VK_RMENU as usize] &= !0x80;
+        }
+        if self.intersects(Self::CAPS_LOCK) {
+            key_state[winuser::VK_CAPITAL as usize] |= 0x01;
+        } else {
+            key_state[winuser::VK_CAPITAL as usize] &= !0x01;
+        }
+    }
+
+    /// Removes the control modifier if the alt modifier is not present.
+    /// This is useful because on Windows: (Control + Alt) == AltGr
+    /// but we don't want to interfere with the AltGr state.
+    pub fn remove_only_ctrl(mut self) -> WindowsModifiers {
+        if !self.contains(WindowsModifiers::ALT) {
+            self.remove(WindowsModifiers::CONTROL);
+        }
+        self
+    }
+}
+
+pub(crate) struct Layout {
+    pub hkl: u64,
+
+    /// Maps a Windows virtual key to a `Key`.
+    ///
+    /// Only those keys are mapped for which the modifier state is not needed when mapping. Making
+    /// this field separate from the `keys` field, saves having to add NumLock as a modifier to
+    /// `WindowsModifiers`, which would double the number of items in keys.
+    pub simple_vkeys: HashMap<c_int, Key<'static>>,
+
+    /// Maps a modifier state to group of key strings
+    /// Not using `ModifiersState` here because that object cannot express caps lock
+    /// but we need to handle caps lock too.
```

Comment:
```suggestion
    /// We're not using `ModifiersState` here because that object cannot express caps lock,
    /// but we need to handle caps lock too.
```